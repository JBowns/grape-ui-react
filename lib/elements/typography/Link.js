'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Link = undefined;

var _templateObject = _taggedTemplateLiteral(['', ''], ['', '']);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _styledComponents = require('styled-components');

var _styledComponents2 = _interopRequireDefault(_styledComponents);

var _textStyles = require('./textStyles');

var _componentHelpers = require('../../utils/componentHelpers');

var _reactRouterDom = require('react-router-dom');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _taggedTemplateLiteral(strings, raw) { return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

var linkPrimitiveFactory = function linkPrimitiveFactory(props, css) {
  var to = props.to,
      otherProps = _objectWithoutProperties(props, ['to']);

  var propsForPrimitive = null;
  var Primitive = null;
  if (to) {
    Primitive = (0, _styledComponents2.default)(_reactRouterDom.Link)(_templateObject, css);
    propsForPrimitive = props;
  } else {
    Primitive = _styledComponents2.default.a(_templateObject, css);
    propsForPrimitive = otherProps;
  }
  return _react2.default.createElement(Primitive, propsForPrimitive);
};

var linkFactory = function linkFactory(props) {
  var color = props.color,
      hoverColor = props.hoverColor,
      otherProps = _objectWithoutProperties(props, ['color', 'hoverColor']);

  var overrides = Object.assign({}, props, {
    color: color || 'brandLink',
    hoverColor: hoverColor || 'brandLinkHover'
  });
  var baseStyles = (0, _textStyles.textStylesBase)(overrides);
  var css = '\n    ' + baseStyles + '\n    &:hover,\n    &:active {\n      color: ' + (0, _componentHelpers.resolveColor)(overrides.hoverColor) + ';\n      text-decoration: none;\n    }\n  ';
  return linkPrimitiveFactory(otherProps, css);
};

linkFactory.propTypes = {
  color: _propTypes2.default.string,
  hoverColor: _propTypes2.default.string
};

linkFactory.defaultProps = {
  color: '',
  hoverColor: ''
};

linkPrimitiveFactory.propTypes = {
  to: _propTypes2.default.string
};

linkPrimitiveFactory.defaultProps = {
  to: ''
};
var Link = function Link(props) {
  return linkFactory(props);
};

exports.Link = Link;