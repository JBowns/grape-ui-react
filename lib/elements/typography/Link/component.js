"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.LinkComponent = undefined;

var _react = require("react");

var _react2 = _interopRequireDefault(_react);

var _propTypes = require("prop-types");

var _propTypes2 = _interopRequireDefault(_propTypes);

var _componentHelpers = require("../../../utils/componentHelpers");

var _reactRouterDom = require("react-router-dom");

var _styledHelpers = require("../../../utils/styledHelpers");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(source, true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(source).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

var LinkComponent = exports.LinkComponent = function LinkComponent(_ref) {
  var children = _ref.children,
      to = _ref.to,
      props = _objectWithoutProperties(_ref, ["children", "to"]);

  var trimmedProps = (0, _componentHelpers.removeSomeProps)(props, Object.keys(_styledHelpers.typography.propTypes));

  if (to) {
    var linkProps = _objectSpread({
      to: to
    }, trimmedProps);

    return _react2["default"].createElement(_reactRouterDom.Link, linkProps, children);
  }

  return _react2["default"].createElement("a", trimmedProps, children);
};

LinkComponent.propTypes = {
  children: _propTypes2["default"].any.isRequired,
  to: _propTypes2["default"].string
};
LinkComponent.defaultProps = {
  to: ''
};