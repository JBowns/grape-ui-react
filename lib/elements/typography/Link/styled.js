"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Link = undefined;

var _propTypes = require("prop-types");

var _propTypes2 = _interopRequireDefault(_propTypes);

var _styledComponents = require("styled-components");

var _styledComponents2 = _interopRequireDefault(_styledComponents);

var _styledSystem = require("styled-system");

var _globalStyles = require("../../../global-styles");

var _styledHelpers = require("../../../utils/styledHelpers");

var _component = require("./component");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _templateObject() {
  var data = _taggedTemplateLiteral(["\n  ", "\n  ", "\n  ", "\n  ", "\n  ", "\n  ", "\n  ", "\n  ", "\n  ", "\n  ", "\n  ", "\n  &:hover,\n  ", "\n  "]);

  _templateObject = function _templateObject() {
    return data;
  };

  return data;
}

function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

var hoverStyle = function hoverStyle(props) {
  return "&:active {\n  color: ".concat((0, _styledHelpers.resolveColor)(props.hoverStyle, (0, _globalStyles.getGlobalOverrides)(props)), ";\n  cursor: pointer;\n}");
};

var Link = (0, _styledComponents2["default"])(_component.LinkComponent)(_templateObject(), _styledHelpers.colorCore, _styledHelpers.ellipsisCore, _styledHelpers.fontFamilyCore, _styledHelpers.fontSizeCore, _styledSystem.fontWeight, _styledHelpers.letterSpacingCore, _styledHelpers.lineHeightCore, _styledHelpers.fontStyleCore, _styledHelpers.textAlignCore, _styledHelpers.textDecorationCore, _styledSystem.space, hoverStyle);
Link.propTypes = _objectSpread({}, _styledHelpers.typography.propTypes, {
  hoverStyle: _propTypes2["default"].string,
  to: _propTypes2["default"].string
});
Link.defaultProps = _objectSpread({}, _styledHelpers.defaultStylesBase, {
  color: 'brandLink',
  hoverStyle: 'brandLinkHover',
  textDecoration: 'none'
});
Link.Router = Link;
exports.Link = Link;