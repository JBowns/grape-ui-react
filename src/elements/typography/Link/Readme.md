When adding in links, you can use `<Link>`. What's cool about this tag is that it will determine if it should use React Router's `<Link>` or `<a>`.

**NOTE:** A Router needs to be defined if using Link the React Router way.

```jsx in Markdown
import { BrowserRouter as Router } from 'react-router-dom';
import { Flex } from 'src/elements/grid'; // ... from 'grape-ui-react'

<Router>
  <Flex
    flexDirection="column"
  >
    <Link to="./">
      {'By using '}
      <code>to</code>
      , it is using React Router.
    </Link>
    <Link href="#">
      {'Leaving out '}
      <code>to</code>
      , it is using a native anchor tag.
    </Link>
    <Link
      href="http://www.google.com"
      target="_blank"
    >
      {'Basic '}
      <code>target</code>
      {' attribute works as well.'}
    </Link>
    <Link
      emailHref={{
        to: 'emailaddress@grapeui.com'
      }}>
      {'For mailto links '}
      <code>
        href=&quot;mailto:emailaddress@grape-ui.com&quot;
      </code>
      .
    </Link>
    <Link
      emailHref={{
        bcc: 'bcc@grapeui.com, bcc2@grapeui.com',
        body: 'WHOAAAAAAAAAAA big body.',
        cc: 'cc@grapeui.com',
        subject: 'Kitchen Sink Example',
      }}
    >
      Kitchen sink email example.
    </Link>
  </Flex>
</Router>
```
