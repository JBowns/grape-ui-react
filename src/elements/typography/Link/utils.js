export function emailHrefString(props) {
  const otherProps = {
    bcc: props.bcc,
    body: props.body,
    cc: props.cc,
    subject: props.subject,
  };
  const trimmedProps = Object.keys(otherProps).filter(key => !!props[key]).map(key => encodeURI(`${key}=${props[key]}`)).join('&');
  return `${props.to || trimmedProps ? 'mailto:' : ''}${props.to || ''}${trimmedProps ? `?${trimmedProps}` : ''}`;
}
