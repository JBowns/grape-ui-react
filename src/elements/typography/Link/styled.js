import PropTypes from 'prop-types';
import styled from 'styled-components';
import { LinkComponent } from './component';
import { fontWeight } from 'styled-system';
import {
  colorCore,
  defaultStylesBase,
  fontFamilyCore,
  fontSizeCore,
  fontStyleCore,
  letterSpacingCore,
  lineHeightCore,
  resolveColor,
  textAlignCore,
  textDecorationCore,
  typography,
} from 'src/utils/styledHelpers';

const hoverColor = props => `&:active {
  color: ${resolveColor(props.hoverColor)};
 }`;

const Link = styled(LinkComponent)`
  ${colorCore}
  ${fontFamilyCore}
  ${fontSizeCore}
  ${fontWeight}
  ${letterSpacingCore}
  ${lineHeightCore}
  ${fontStyleCore}
  ${textAlignCore}
  ${textDecorationCore}
  &:hover,
  ${hoverColor}
  `;

Link.propTypes = {
  ...typography.propTypes,
  hoverColor: PropTypes.string,
  to: PropTypes.string,
};

Link.defaultProps = {
  ...defaultStylesBase,
  color: 'brandLink',
  hoverColor: 'brandLinkHover',
  textDecoration: 'none',
};

Link.Router = Link;
export { Link };
