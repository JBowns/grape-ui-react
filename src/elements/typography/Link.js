import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { passThrough, resolveColor } from 'src/utils/componentHelpers';
import { Link as ReactRouterLink } from 'react-router-dom';

import {
  getFontFamily,
  getFontSize,
  getFontWeight,
  getLetterSpacing,
  getLineHeight,
  getFontStyle,
  getColor,
  getTextAlign,
  getTextDecoration,
  typography,
} from './textStyles';

const getHoverColor = props => `&:active {
  color: ${resolveColor(props.hoverColor)};
 }`;

const LinkComponent = ({
  children, className, to, ...props
}) => {
  if (to) {
    const linkProps = {
      ...passThrough(LinkComponent, props),
      to,
    };
    return (
      <ReactRouterLink {...linkProps} className={className}>
        {children}
      </ReactRouterLink>
    );
  }
  return (
    <a {...passThrough(LinkComponent, props)} className={className}>
      {children}
    </a>
  );
};

LinkComponent.propTypes = {
  children: PropTypes.any.isRequired,
  className: PropTypes.string,
  hoverColor: PropTypes.string,
  ...typography.propTypes,
  to: PropTypes.string,
};

LinkComponent.defaultProps = {
  className: '',
  hoverColor: '',
  to: '',
};
const Link = styled(LinkComponent)`
  ${getFontFamily}
  ${getFontSize}
  ${getFontWeight}
  ${getLetterSpacing}
  ${getLineHeight}
  ${getFontStyle}
  ${getColor}
  ${getTextAlign}
  ${getTextDecoration}
  &:hover,
  ${getHoverColor}
  `;

Link.propTypes = {
  ...typography.propTypes,
  hoverColor: PropTypes.string,
  to: PropTypes.string,
};

Link.defaultProps = {
  color: 'brandLink',
  hoverColor: 'brandLinkHover',
  textDecoration: 'none',
};

Link.Router = Link;
export { Link };
