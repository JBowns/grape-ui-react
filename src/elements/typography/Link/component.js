import React from 'react';
import PropTypes from 'prop-types';
import { removeSomeProps } from 'src/utils/componentHelpers';
import { Link as ReactRouterLink } from 'react-router-dom';
import { typography } from 'src/utils/styledHelpers';

export const LinkComponent = ({
  children, to, ...props
}) => {
  const trimmedProps = removeSomeProps(props, Object.keys(typography.propTypes));
  if (to) {
    const linkProps = {
      to,
      ...trimmedProps,
    };
    return (
      <ReactRouterLink {...linkProps}>
        {children}
      </ReactRouterLink>
    );
  }
  return (
    <a {...trimmedProps}>
      {children}
    </a>
  );
};

LinkComponent.propTypes = {
  children: PropTypes.any.isRequired,
  to: PropTypes.string,
};

LinkComponent.defaultProps = {
  to: '',
};
