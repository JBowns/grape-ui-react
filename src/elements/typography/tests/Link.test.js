import React from 'react';
import { Link } from '../Link';
import renderer from 'react-test-renderer';
import { MemoryRouter as Router } from 'react-router-dom';
import 'jest-styled-components';

describe('Link Component', () => {
  it('should return an anchor tag', () => {
    const component = renderer.create(<Link>Lorem Ipsum</Link>);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('should return a react router tag', () => {
    const component = renderer.create(<Router><Link to="./home">Lorem Ipsum</Link></Router>);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
