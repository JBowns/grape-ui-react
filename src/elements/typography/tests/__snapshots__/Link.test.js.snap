// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link Component should return a react router tag 1`] = `
.c0 {
  font-size: calc(1rem * 1);
  font-weight: inherit;
  line-height: 1.5;
  color: hsl(186.8,100%,41.6%);
  text-align: inherit;
}

.c0:hover,
.c0:active {
  color: hsl(207,90%,54%);
  -webkit-text-decoration: none;
  text-decoration: none;
}

<a
  className="c0"
  href="./home"
  onClick={[Function]}
>
  Lorem Ipsum
</a>
`;

exports[`Link Component should return an anchor tag 1`] = `
.c0 {
  font-size: calc(1rem * 1);
  font-weight: inherit;
  line-height: 1.5;
  color: hsl(186.8,100%,41.6%);
  text-align: inherit;
}

.c0:hover,
.c0:active {
  color: hsl(207,90%,54%);
  -webkit-text-decoration: none;
  text-decoration: none;
}

<a
  className="c0"
>
  Lorem Ipsum
</a>
`;
